sdk/eventhubs/azure-messaging-eventhubs/inc/azure/messaging/eventhubs/models/event_data.hpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#pragma once
+#include <azure/core/amqp.hpp>
+namespace Azure { namespace Messaging { namespace EventHubs { namespace Models {
+
+  /** @brief The type of the body of an EventData Message.
+   *
+   */
+  struct EventDataBody
+  {
+    /** @brief Value is encoded / decoded as the amqp - value section in the body.
+     *
+     * The type of Value can be any of the AMQP simple types, as listed in the comment for
+     * AmqpMessage, as well as slices or maps of AMQP simple types.
+     */
+    Azure::Core::Amqp::Models::AmqpValue Value;
+
+    /** @brief  Sequence is encoded/decoded as one or more amqp-sequence sections in the body.
+     *
+     * The values of the slices are are restricted to AMQP simple types, as listed in the comment
+     * for AmqpMessage.
+     */
+    Azure::Core::Amqp::Models::AmqpList Sequence;
+
+    /** @brief Data is encoded decoded as multiple data sections in the body.
+     */
+    Azure::Core::Amqp::Models::AmqpBinaryData Data;
+  };
+
+  /** @brief Represents an event sent to the Azure Event Hubs service.
+   */
+  struct EventData
```

Comment:
It is likely you want to do something around the operator= and copy constructor, because slicing is possible and it may not be expected, the safest route is to delete them.

Other option - ReceivedEventData does not derive, but duplicates, or ReceivedEventData includes EventData as its field.

If you want ReceivedEventData to have operator= and a copy constructor, declare EventData's operator= and a copy constructor as `= default` but `protected`, and then declare them in ReceivedEventData and call them explicitly.

Reply:
Interestingly enough, I don't mind slicing. You *should* be able to pass a ReceivedEventData into anything that takes an EventData and all the extra information in the ReceivedEventData should not be recoverable, since it is information about when the EventData was received (and should never be passed on).